Introduce registry of XML models

jonseymour opened the pull request:
This series introduces a registry of XML models.

The XML models contain metadata that is required to successfully map a stream of bytes onto a Block{N} or Block{N}Repeat structure and vice versa and to apply scaling factors as and when required.

An alternative to the current placement of the generated logic for the model registrations would be to run a separate generator that creates a models.go file in the smdx package.

jonseymour committed:
Id has to be a uint16, so reflect this in the model.

jonseymour committed:
Add support for a registry of models.

Signed-off-by: Jon Seymour <jon@wildducktheories.com>

jonseymour committed:
Add support for generating model objects from the XML representation.

jonseymour committed:
Result of regenerating with model objects.

jonseymour commented:
Thinking about this some more...

It is probably uncool to load the entire Sunspec model space into every Go executable particularly if you are only intend to use a couple of models in a given context. One way to fix this would be to put each model registration into its own go package along with constants that name each point, that way the Go compiler can remove models that are not actually used at compile time.

Also, it isn't clear to me that defining Block{N} structures is really that useful in practice for the following reasons. While it *might* be reasonable to read an entire model into memory, it is almost never correct to write a whole model and it isn't clear how partial updates written directly into a Block{N} structure would be  intercepted and converted into Modbus write operations. Further the Go memory layout will not be byte-for-byte identical with the wire format of a block because of byte ordering issues and the way that strings are (currently) represented in the Block{N} structs. This means that there is no technical advantage to defining a contiguous block structure, at least on the Modbus side.

It seems to me that a better way for a programmer to interact with a model would be to specify, with symbolic identifiers, the set of points to be read, read the registers corresponding to those points into a buffer, and then use the point identifier with strongly typed accessor methods to read and write the points of interest to/from the buffer of interest.

So: 

        Block b;

        b.Prepare(model1.Mn, model1.DA)
        b.Refresh()  // executes a sequence of ReadHoldingRegisters calls to populate a buffer
        da := b.GetUint16(model1.DA) // unpacks the buffer

Similarly for write:

        b.PutUint16(model1.DA, 1) // write into a local buffer
        b.Flush() // flush the buffer into WriteHoldingRegisters calls

So, what I plan to do is this:

 - change the generator to generate one package under core/models for each model
 - in the model specific package, define constants for each point name
 - in the model specific package, register the XML model
 - ~~don't generate a Block{N} or Block{N}Repeat struct for each model~~
 - define an API for reading and writing block buffers from an underlying Modbus API

**Updated:** thinking about it some more, I guess the Block types might be useful place for an application programmer to store partial reads that return multiple points even if full reads and writes of a block aren't always practical in themselves. So I have kept the block types around in the reorganised generation.

jonseymour committed:
Add a DoModels feature to iterate over all the models.

jonseymour committed:
Add a tool for checking the XML models for consistency.

jonseymour committed:
reorganisation: use a single template and single template execution.

Also, eliminate separate model variable and inline the contents of the
initializer into a RegisterModel call within the body of the init() function.

Signed-off-by: Jon Seymour <jon@wildducktheories.com>  t

jonseymour committed:
Write each model into its own Go package, and ...

- generate a loader file to allow all models to be loaded at the callers' option (see checkmodels).
- add string constants for the point names in each model

Signed-off-by: Jon Seymour <jon@wildducktheories.com>

jonseymour commented:
I am wondering if the models might better be generated within packages named like "sunspec/model{N}" or "sunspec/models/model{N}" rather than "core/models/model{N}"?

jonseymour committed:
Use constants for typelabels rather than literals.

jonseymour committed:
Regenerate model with type labels.

crabmusket commented on generators/core.go:
```diff
@@ -10,81 +10,97 @@ import (
+	Name: "{{$model.Name}}",
+	Length: {{$model.Length}},
+	Blocks: []smdx.BlockElement{ {{range $model.Blocks}}
 			smdx.BlockElement{ {{if gt (len .Name) 0 }}Name: "{{.Name}}",{{end}}
```
Not that it's a huge deal, but the output with `BlockElement{Name: ""` is a bit odd

jonseymour replied:
The construction above should explicitly prevent `BlockElement{Name: ""` while allowing `BlockElement{Name: "foobar"`. 

That said, I think it is quite possible I didn't understand the actual meaning of your comment.

crabmusket commented:
* I agree with all your points about the `Block{N}` structs. The original intent of those types was for use in application-side code (e.g. analysis of the data stream in a monitoring device) but that can be accomplished using your buffer model.
* The namespace is nested under `github.com/crabmusket/gosunspec/`, so I'm not a fan of repeating `sunspec`, but `core` isn't very descriptive so a change might be for the best.

This looks good! Is it mergeable? Given I'm so distant from the code, I'll basically merge anything that works for you. But I will try to at least give a cursory review of everything before I do!

jonseymour committed:
Add sunspec specifications as submodule.

By speifying a particular submodule commit, you can guarantee
that the code generation will be the same, irrespective of whether
the upstream specification has since changed. This would be useful
in cases where something in an updated specification violates
an assumption made code generation code.

Signed-off-by: Jon Seymour <jon@wildducktheories.com>

jonseymour committed:
Reorganise layout of models:

models/ -> spec/
core/models/ -> models/
generators/core.go -> generators/models.go

Signed-off-by: Jon Seymour <jon@wildducktheories.com>

jonseymour committed:
Regenerate model in new location (core/models -> models)

Signed-off-by: Jon Seymour <jon@wildducktheories.com>

jonseymour commented:
Ok, so what I have done with these last 3 commits, is reorganise the namespace so that the models are now under `models/ `rather than `core/models/`.

To make this work, I have to put the SMDX modules into `spec/`, to remove the conflict that would otherwise arise.

I have also made `spec/` a submodule which has the advantage of allowing the particular spec used to generate a model tree to be tied exactly to the model tree it generates.

Subsequent commits will fix the compilation errors in`xml` which I hadn't noticed until now.

jonseymour committed:
Fixup compilation errors and make tests consistent.

This code will ultimately either need to be generated or made reflective.

Signed-off-by: Jon Seymour <jon@wildducktheories.com>

jonseymour committed:
Just use the golang string type for the sunspec string type.

jonseymour committed:
Regenerate the model to use golang strings.

Signed-off-by: Jon Seymour <jon@wildducktheories.com>

jonseymour committed:
Remove generated go binary from git repo.

Signed-off-by: Jon Seymour <jon@wildducktheories.com>

jonseymour commented:
I decided to remove the core.String type and replace all uses of it with the golang string type based on ease of use considerations.

crabmusket commented:
Seems fine. `core.String` was one of those things I wasn't sure about. I think there might have been a couple of other types that I didn't alias to core types as well.

I like the idea of making the spec a submodule, good thinking.

Let me know when this is mergeable :)

jonseymour commented:
It is ready to merge in its current state.

Something that might be worth doing is moving the core/* types into the top level, then changing the package from core to "sunspec". That way everyone gets the sunspec prefix by default. If you'd like me to do that, I can do that next.

I am currently working on an interface model that should be general enough to allow working with models that are backed by XML documents or by a modbus interface, but I won't push that until these changes have been merged.

crabmusket commented:
Prefix by default would probably be good :+1:
